usr.bin/su/su.c

```diff
@@ -0,0 +1,557 @@
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the University nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/cdefs.h>
+#include <sys/param.h>
+#include <sys/time.h>
+#include <sys/resource.h>
+#include <err.h>
+#include <errno.h>
+#include <grp.h>
+#include <paths.h>
+#include <pwd.h>
+#include <stdio.h>
+#ifdef SKEY
+#include <skey.h>
+#endif
+#include <stdlib.h>
+#include <string.h>
+#include <syslog.h>
+#include <time.h>
+#include <tzfile.h>
+#include <unistd.h>
+#include <util.h>
+
+#ifdef LOGIN_CAP
+#include <login_cap.h>
+#endif
+
+#ifdef KERBEROS5
+#include <krb5.h>
+#endif
+
+#ifdef ALLOW_GROUP_CHANGE
+#include "grutil.h"
+#endif
+#include "suutil.h"
+
+#ifdef KERBEROS5
+#define	ARGSTRX	"-Kdflm"
+static int kerberos5(char *, const char *, uid_t);
+int use_kerberos = 1;
+#else
+#define	ARGSTRX	"-dflm"
+#endif
+
+#ifndef	SU_GROUP
+#define	SU_GROUP	"wheel"
+#endif
+
+#define GROUP_PASSWORD	"Group Password:"
+
+#ifdef LOGIN_CAP
+#define ARGSTR	ARGSTRX "c:"
+#else
+#define ARGSTR ARGSTRX
+#endif
+
+static int check_ingroup(int, const char *, const char *, int);
+
+int
+main(int argc, char **argv)
+{
+	extern char **environ;
+	struct passwd *pwd;
+	char *p;
+	uid_t ruid;
+	int asme, ch, asthem, fastlogin, prio, gohome;
+	enum { UNSET, YES, NO } iscsh = UNSET;
+	const char *user, *shell, *avshell;
+	char *username, **np;
+#ifdef SU_ROOTAUTH
+	char *userpass;
+#endif
+	char *class;
+	char shellbuf[MAXPATHLEN], avshellbuf[MAXPATHLEN];
+#ifdef LOGIN_CAP
+	time_t pw_warntime = _PASSWORD_WARNDAYS * SECSPERDAY;
```

Comment:
This line was moved here (between macro) because it is only used when `LOGIN_CAP` is used.